OrbitGl/TimerChain.h

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_GL_TIMER_CHAIN_
+#define ORBIT_GL_TIMER_CHAIN_
+
+#include <assert.h>
+
+#include <algorithm>
+#include <atomic>
+#include <cstdint>
+#include <limits>
+
+#include "TextBox.h"
+
+static constexpr int kBlockSize = 1024;
+class TimerChain;
+
+// TimerBlock is a straightforward specialization of Block (see BlockChain.h)
+// with the added bonus that it keeps track of the minimum and maximum
+// timestamps of all timers added to it. This allows trivial rejection of an
+// entire block by using the Intersects(t_min, t_max) method. This effectively
+// tests if any of the timers stored in this block intersects with the [t_min,
+// t_max] interval.
+class TimerBlock {
+  friend class TimerChain;
+  friend class TimerChainIterator;
+
+ public:
+  TimerBlock(TimerChain* chain, TimerBlock* prev)
+      : prev_(prev), next_(nullptr), chain_(chain), size_(0) {}
+
+  ~TimerBlock() {}
+
+  // Adds an item to the block. If capacity of this block is reached, a new
+  // blocked is allocated and the item is added to the new block.
+  void Add(const TextBox& item);
+
+  // Tests if [min, max] intersects with [min_timestamp, max_timestamp], where
+  // {min, max}_timestamp are the minimum and maximum timestamp of the timers
+  // that have so far been added to this block.
+  bool Intersects(uint64_t min, uint64_t max);
+
+  uint32_t size() const { return size_; }
+
+  TextBox& operator[](std::size_t idx) { return data_[idx]; }
+
+  const TextBox& operator[](std::size_t idx) const { return data_[idx]; }
+
+ private:
+  TimerBlock* prev_;
+  TimerBlock* next_;
+  TimerChain* chain_;
+  uint32_t size_;
+  TextBox data_[kBlockSize];
+
+  uint64_t min_timestamp_;
```

Comment:
Are these initialized?

Reply:
This was indeed a bug, thanks for spotting this.